### PR TITLE
WIP Registration audit fixes

### PR DESCRIPTION
**Pull request by richardliang**

### CODE REVIEW

**Comment by 0xSachinK** [comment]
Solves https://github.com/zkp2p/zk-p2p/issues/113

Instead of taking the `venmo_actor_id_idx` as input from the user. Calculate it inside the circuit.

This code is generic and can be used in every regex circuit.
```javascript
    signal output reveal_start_idx;
    signal prefix_sum[num_bytes+1];
    signal temp_index[num_bytes+1];
    component is_eq_one[num_bytes];
    prefix_sum[0] <== 0;
    temp_index[0] <== 0;
    for (var i = 1; i <= num_bytes; i++) {
        prefix_sum[i] <== prefix_sum[i-1] + (1 - states[i][1]);

        is_eq_one[i-1] = IsEqual();
        is_eq_one[i-1].in[0] <== prefix_sum[i];
        is_eq_one[i-1].in[1] <== 1;
        temp_index[i] <== temp_index[i-1] + is_eq_one[i-1].out * (i-1);
    }

    reveal_start_idx <== temp_index[num_bytes];
 ```

### Explanation
The `states[i][1]` array is 1 at all the values that match the regex. Initialize a prefix sum array that calculates prefix sum of the `states[i][1]` array. The Prefix sum array has 1 only at the beginning index of the reveal array (`venmo_actor_id` in this case). Iterate over the prefix sum array and compare each value to a 1, resulting in a mask array that has 1 only at the beginning index of the reveal array and 0 elsewhere. Perform a dot product of the mask array with an array storing all indices, and we get the reveal_start_idx output.

File: circuits-circom/regexes/venmo_actor_id.circom [hunk 1]
```diff
@@ -189,4 +189,21 @@ template VenmoActorId (msg_bytes) {
...
     }
+
+    signal output reveal_start_idx;
+    signal prefix_sum[num_bytes+1];
+    signal temp_index[num_bytes+1];
+    component is_eq_one[num_bytes];
+    prefix_sum[0] <== 0;
+    temp_index[0] <== 0;
+    for (var i = 1; i <= num_bytes; i++) {
+        prefix_sum[i] <== prefix_sum[i-1] + (1 - states[i][1]);
```
Comment (kcharbo3) [inline circuits-circom/regexes/venmo_actor_id.circom:200]: Should `1 - states[i][1]` be changed to `1 * states[i][1]` ? 

Also I think we can get rid of temp_index[num_bytes+1] and instead just do `reveal_start_idx <== prefix_sum[num_bytes]` right? 

I'm thinking if states[...][1] = 000011100 for example, then prefix_sum[...] = 000012333. So we just want to return the last value of prefix sum. Let me know if my thinking is correct.

Comment (kcharbo3) [inline circuits-circom/regexes/venmo_actor_id.circom:200]: Should `(1 - states[i][1])` be `(1 * states[i][1])` instead? That way prefix_sum is only 1 at the first nonzero states value.

Comment (richardliang) [reply circuits-circom/regexes/venmo_actor_id.circom:200]: fixed here https://github.com/zkp2p/zk-p2p/pull/119/commits/f0062fce89cf44fd63aa33c12856bf96c286f942

**Pull request merged**